apollo-router/src/spec/query/change.rs

```diff
@@ -92,198 +149,335 @@ impl<'a> QueryHashVisitor<'a> {
             schema::Type::NonNullNamed(name) => {
                 "!".hash(self);
-                self.hash_type_by_name(name.as_str())
+                self.hash_type_by_name(name.as_str())?;
             }
             schema::Type::List(t) => {
                 "[]".hash(self);
-                self.hash_type(t)
+                self.hash_type(t)?;
             }
             schema::Type::NonNullList(t) => {
                 "[]!".hash(self);
-                self.hash_type(t)
+                self.hash_type(t)?;
             }
         }
+        "^TYPE-END".hash(self);
+        Ok(())
     }
 
     fn hash_field(
         &mut self,
-        parent_type: String,
-        type_name: String,
+        parent_type: &str,
         field_def: &FieldDefinition,
-        arguments: &[Node<Argument>],
+        node: &executable::Field,
     ) -> Result<(), BoxError> {
-        if self.hashed_fields.insert((parent_type.clone(), type_name)) {
-            self.hash_type_by_name(&parent_type)?;
+        "^FIELD".hash(self);
+        self.hash_field_definition(parent_type, field_def)?;
 
-            field_def.name.hash(self);
+        "^ARGUMENT_LIST".hash(self);
+        for argument in &node.arguments {
+            self.hash_argument(argument);
+        }
+        "^ARGUMENT_LIST_END".hash(self);
 
-            for argument in &field_def.arguments {
-                self.hash_input_value_definition(argument)?;
-            }
+        self.hash_directive_list_ast(&node.directives);
 
-            for argument in arguments {
-                self.hash_argument(argument);
-            }
+        node.alias.hash(self);
+        "^FIELD-END".hash(self);
 
-            self.hash_type(&field_def.ty)?;
+        Ok(())
+    }
 
-            for directive in &field_def.directives {
-                self.hash_directive(directive);
-            }
+    fn hash_field_definition(
+        &mut self,
+        parent_type: &str,
+        field_def: &FieldDefinition,
+    ) -> Result<(), BoxError> {
+        "^FIELD_DEFINITION".hash(self);
+
+        let field_index = (parent_type.to_string(), field_def.name.as_str().to_string());
+        if self.hashed_field_definitions.contains(&field_index) {
+            return Ok(());
+        }
+
+        self.hashed_field_definitions.insert(field_index);
+
+        self.hash_type_by_name(parent_type)?;
 
-            self.hash_join_field(&parent_type, &field_def.directives)?;
+        field_def.name.hash(self);
+        self.hash_type(&field_def.ty)?;
+
+        // for every field, we also need to look at fields defined in `@requires` because
+        // they will affect the query plan
+        self.hash_join_field(parent_type, &field_def.directives)?;
+
+        self.hash_directive_list_ast(&field_def.directives);
+
+        "^ARGUMENT_DEF_LIST".hash(self);
+        for argument in &field_def.arguments {
+            self.hash_input_value_definition(argument)?;
```

Comment:
I'm not 100% sure, but it looks like it also should hash `@fromContext`:
https://www.apollographql.com/docs/graphos/reference/federation/directives#fromcontext
I looked into `apollo-federation,` and it seems that Rust QP doesn't support it but it is implemented only in JS QP.
I don't think this directive will affect entity cache because you also hashing selection set of subgraphql query (I'm just assuming here)

However, this directive should affect QP (at least for the JS planner), breaking QP caching.